ethcore/src/snapshot/account.rs

```diff
@@ -62,53 +63,50 @@ impl CodeState {
 
 // walk the account's storage trie, returning an RLP item containing the
 // account properties and the storage.
-pub fn to_fat_rlp(acc: &BasicAccount, acct_db: &AccountDB, used_code: &mut HashSet<H256>) -> Result<Bytes, Error> {
+pub fn to_fat_rlps(acc: &BasicAccount, acct_db: &AccountDB, used_code: &mut HashSet<H256>, preferred_size: usize) -> Result<Vec<Bytes>, Error> {
+	const AVERAGE_BYTES_PER_STORAGE_ENTRY: usize = 47;
 	if acc == &ACC_EMPTY {
-		return Ok(::rlp::NULL_RLP.to_vec());
+		return Ok(vec![::rlp::NULL_RLP.to_vec()]);
 	}
 
 	let db = TrieDB::new(acct_db, &acc.storage_root)?;
 
-	let mut pairs = Vec::new();
+	let pair_chunks: Vec<Vec<_>> = db.iter()?.chunks(preferred_size / AVERAGE_BYTES_PER_STORAGE_ENTRY).into_iter().map(|chunk| chunk.collect()).collect();
+	pair_chunks.into_iter().pad_using(1, |_| Vec::new(), ).map(|pairs| {
+		let mut stream = RlpStream::new_list(pairs.len());
 
-	for item in db.iter()? {
-		let (k, v) = item?;
-		pairs.push((k, v));
-	}
-
-	let mut stream = RlpStream::new_list(pairs.len());
-
-	for (k, v) in pairs {
-		stream.begin_list(2).append(&k).append(&&*v);
-	}
+		for r in pairs {
+			let (k, v) = r?;
+			stream.begin_list(2).append(&k).append(&&*v);
+		}
 
-	let pairs_rlp = stream.out();
-
-	let mut account_stream = RlpStream::new_list(5);
-	account_stream.append(&acc.nonce)
-				  .append(&acc.balance);
-
-	// [has_code, code_hash].
-	if acc.code_hash == SHA3_EMPTY {
-		account_stream.append(&CodeState::Empty.raw()).append_empty_data();
-	} else if used_code.contains(&acc.code_hash) {
-		account_stream.append(&CodeState::Hash.raw()).append(&acc.code_hash);
-	} else {
-		match acct_db.get(&acc.code_hash) {
-			Some(c) => {
-				used_code.insert(acc.code_hash.clone());
-				account_stream.append(&CodeState::Inline.raw()).append(&&*c);
-			}
-			None => {
-				warn!("code lookup failed during snapshot");
-				account_stream.append(&false).append_empty_data();
+		let pairs_rlp = stream.out();
+
+		let mut account_stream = RlpStream::new_list(5);
+		account_stream.append(&acc.nonce)
+					  .append(&acc.balance);
+
+		// [has_code, code_hash].
+		if acc.code_hash == SHA3_EMPTY {
```

Comment:
spec states that the hash may only be given when an account with a SHA3 hash strictly less has the same code. that's not (necessarily) the case here, so the spec implies it may be the case that the full code should be repeated.